test/store_unittest.cc

```diff
@@ -0,0 +1,288 @@
+}
+
+TEST_F(MetadataAgentStoreTest, UpdateMetadataEntryStore) {
+  SetupStore();
+  MonitoredResource mr("some_resource", {});
+  MetadataStore::Metadata m1(
+      "default-version1",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::Parser::FromString("{\"f\":\"hello\"}")
+      );
+  store->UpdateMetadata(mr, std::move(m1));
+  std::map<MonitoredResource, MetadataStore::Metadata> metadata_map =
+      GetMetadataMap();
+  EXPECT_FALSE(metadata_map.empty());
+  EXPECT_EQ("default-version1", metadata_map.at(mr).version);
+  MetadataStore::Metadata m2(
+      "default-version2",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::Parser::FromString("{\"f\":\"hello\"}")
+      );
+  store->UpdateMetadata(mr, std::move(m2));
+  metadata_map = GetMetadataMap();
+  EXPECT_EQ("default-version2", metadata_map.at(mr).version);
+}
+
+TEST_F(MetadataAgentStoreTest, PurgeDeletedEntriesStore) {
+  SetupStore();
+  MonitoredResource mr1("some_resource1", {});
+  MonitoredResource mr2("some_resource2", {});
+  MetadataStore::Metadata m1(
+      "default-version1",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::Parser::FromString("{\"f\":\"hello\"}")
+      );
+  MetadataStore::Metadata m2(
+      "default-version2",
+      true,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::Parser::FromString("{\"f\":\"hello\"}")
+      );
+  store->UpdateMetadata(mr1, std::move(m1));
+  store->UpdateMetadata(mr2, std::move(m2));
+  std::map<MonitoredResource, MetadataStore::Metadata> metadata_map =
+      GetMetadataMap();
+  EXPECT_FALSE(metadata_map.empty());
+  EXPECT_EQ("default-version1", metadata_map.at(mr1).version);
+  EXPECT_EQ("default-version2", metadata_map.at(mr2).version);
+  PurgeDeletedEntries();
+  metadata_map = GetMetadataMap();
+  EXPECT_EQ("default-version1", metadata_map.at(mr1).version);
+  EXPECT_THROW(metadata_map.at(mr2), std::out_of_range);
+}
+
+TEST(MetadataTest, ParseMetadata) {
+  MetadataStore::Metadata m(
+      "default-version",
+      false,
+      time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+      time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+      json::Parser::FromString("{\"f\":\"hello\"}")
+      );
+  EXPECT_EQ("default-version", m.version);
+  EXPECT_FALSE(m.is_deleted);
+  EXPECT_FALSE(m.ignore);
+  EXPECT_EQ(time::rfc3339::FromString("2018-03-03T01:23:45.678901234Z"),
+            m.created_at);
+  EXPECT_EQ(time::rfc3339::FromString("2018-03-03T01:32:45.678901234Z"),
+            m.collected_at);
+  const auto& obj = m.metadata->As<json::Object>();
```

Comment:
Careful -- the return value isn't a reference. But here `auto` doesn't buy you anything anyway -- just use `const json::Object* obj`.

Reply:
Done.